### DISCUSSION TITLE
Using without notebooks

### DISCUSSION THREAD

**Discussion by sterisa** [sassoftware/saspy#604] [Q&A]
Hi
Can saspy be used without using notebooks? I seem to get html responses on my calls.
Thanks
Stefano

**Comment by tomweber-sas** [comment]
Yes, of course. You can use it with notebooks, any other UI (vscode, spyder, .....), no UI just line mode in a terminal, or even in python batch scripts. The `results` configuration key/SASsession attribute (that you can change dynamically), controls how tabular results are returned. They can be either html, Pandas data frames, or text. For plots/charts/graphs from analytical methods, those will always be html; can't get then different than that. If you're in a terminal session, you can't display that, but you can set `batch` mode and the html will be returned to you as a string which you can write to a file to then display using some ui that can display html; a browser or whatever you like. So it depends upon the environment you're running in. Sounds like you're running in a way that can't render html?

**Reply by sterisa** [reply]
Re opening this conversation after a long time, sorry
My organization is moving more and more to python and snowflake.
At the same time in my 25 years of sas i always felt uncomfortable with macros.
Long story short i have a lot overnight batches. And i want to continue running the heavy lifting in sas, but manage the code flow from python, essentially replacing macros and making my code more organization friendly.
Is that a sensible usage of saspy? Suggestions on where to start?
I also don't fully understand the philosophy of saspy. I just want to build statements in python and submit them in sas like a macro would do, move a little data between processes (proc contents, proc datasets, warnings and errors), but feels like i may be missing many features.

**Comment by tomweber-sas** [comment]
I can't stand our SAS Macro language either! :)   So, what you're describing is a perfectly fine use of SASPy. You can use it for whatever you want to do. It's simply a Python interface to SAS, so you can integrate SAS into your Python workflow, just use SAS as a data source pulling data into Python and doing everything in python, use it to easily script SAS code execution in a easy language - Python (what you're suggesting) or write other applications in Python which use saspy under the covers, ...  Customers use saspy in a lot of different ways for different reasons. Replacing SAS Macro with Python as your scripting/control language for running SAS code is perfectly reasonable.

In fact, one of the uses of proc python (only available in viya), is that same idea, using it in SAS as a scripting language instead of macro; that's what proc lua was originally written for; to use instead of macro. Proc Python supports more full integration of Python into you SAS workflow than what lua does, but even just using it for scripting SAS code execution is what some users will use it for, though a lot is, of course, data science workflows which are a perfect integration point for SAS and Python. So with SASPy and Proc Python, you can integrate one language into the other.

Perhaps an easy place to start is to look over (download/run/play with) the main notebook in [saspy-examples](https://github.com/sassoftware/saspy-examples/blob/main/SAS_contrib/saspy_example_github.ipynb) to become familiar with the functionality it provides. That goes over most of the code functionality saspy has, in an easy to read and see notebook.  And, also of course, the doc! The [API](https://sassoftware.github.io/saspy/api.html) will help too!

**Comment by srisa-work** [comment]
Does this override my default shell (csh)? I suspect my .cshrc is not getting executed

I see this works in my sas bacthes+interactive sessions:
proc sql;
connect to odbc as snf (dsn=&snf_dsn. user=&snf_usr. password="&snf_pwd");

But this does not:
sas = saspy.SASsession(results="Pandas")
res = sas.submit("proc sql;")
res = sas.submit('connect to odbc as snf (dsn=&snf_dsn. user=&snf_usr. password="&snf_pwd");')
print(res['LOG'])

27    ods listing close;ods html5 (id=saspy_internal) file=stdout options(bitmap_mode='inline') device=svg style=HTMLBlue; ods
27  ! graphics on / outputfmt=png;
NOTE: Writing HTML5(SASPY_INTERNAL) Body file: STDOUT
28    
29    connect to odbc as snf
29  ! (dsn=&snf_dsn. user=&snf_usr. password="&snf_pwd");
ERROR: Could not load /base/sas/SAS_9.4/SASFoundation/9.4/sasexe/sasodb (80 images loaded)
ERROR: /appl/toolbox/linux/sybase/IQ-16_1/lib64/libodbc.so: version `VERS_3.52' not found (required by 
       /base/sas/SAS_9.4/SASFoundation/9.4/sasexe/sasodb)
ERROR: The SAS/ACCESS Interface to ODBC cannot be loaded. The SASODB   code appendage could not be loaded.
ERROR: A Connection to the odbc DBMS is not currently supported, or is not installed at your site.
NOTE: PROC SQL set option NOEXEC and will continue to check the syntax of statements.
30    
31    ods html5 (id=saspy_internal) close;ods listing;

**Reply by tomweber-sas** [reply]
That error is from SAS itself. Given the question, I'm guessing you have something that keeps this error from happening in your shell script? I don't know where SAS is running, or how, or which access method you're using or anything I would need to know to answer this. Where is your client (linux it would seem) but on the same machine as where SAS is? Where is SAS and is it a standalone deployment or a production one (Woprkspace server); it also appears linux, so there's that? What access method are you using to connect? It depends upon these. The only case where your shell script would come into play would be a standalone linux deployment using the STDIO access method with python on the same machine as SAS. If that's the case and some environment variables aren't propagating to the SAS process, then setting them in SAS configuration or at runtime migh be needed. But again, what is it you have in your script that makes this work? Then I might have a better idea.

**Reply by srisa-work** [reply]
SAS runs on Linux,. I am connecting through ssh from Windows.
In both cases the commands are executed on a fresh session.
My ODBC config is in my .cshrc
I think the difference is that a new csh session will automatically run the .cshrc, but apparently an ssh will not
How is this normally handled? Is there some way to source it? Maybve it could be added as an option to add to the ssh statement

**Reply by tomweber-sas** [reply]
ok, that helps a bunch. So, since it's not getting it from that file, it may be that it's not really starting a shell like that. I would have to play with this some to figure out just how that's working. On my system, I have a .profile as well as my .bashrc, so what if you put the odbc config stuff in there instead? Does that work for all cases? Assuming you're connection over ssh as you, then I would think your .profile would get sourced even if the shell file doesn't - is what I'm thinking.

**Reply by tomweber-sas** [reply]
If not, and it's just environment variables you're assigning, then you can assign them with SAS code before trying to use the engine. That may solve it too, if the profile doesn't. Again, what kind of config is it you're counting on in that file?

**Reply by srisa-work** [reply]
The .cshrc does a lot of things (including calling an organization-wide .cshrc)... so ideally I would source the whole file
Absent that, I think the relevant changes are setting some enviroment variables (LD_LIBRARY_PATH + ODBCINI, ODBCSYSINI, ODBCINSTINI)
Never heard about .profile :). checking on that

**Reply by srisa-work** [reply]
looks like .profile is bash specific

**Reply by srisa-work** [reply]
I created a runsas.sh that sources the .cshrc and it worked. I suspect the issue was related to something origanization-specific

**Reply by tomweber-sas** [reply]
Good deal. There is more than one way to do that, so good. As for .profile, it isn't specific to bash, I had is decades ago when using korn shell; .kshrc. But, glad you got your config working!

**Comment by tomweber-sas** [comment]
I tried to reply to the first question, but it seems to be removed such that I couldn't send this reply. But it has info about what you were asking, so I'll just post it here.


There's an example notebook [here](https://github.com/sassoftware/saspy-examples/blob/main/SAS_contrib/saspy_example_github.ipynb) that walks you though all of the general functionality of SASPy. It's a notebook cuz those are useful for examples like this. At cell 63 it shows how to run saspy in a batch python script. Maybe that's something you're looking for? The 'batch' setting changes methods that otherwise would render output (which is expected when running interactively), to return that output instead. In this way you can choose what to do with it yourself; render it, write it to a file so you can open it in another application that can view it (HTML for example). 

I'm not completely sure what scenario you're trying to accomplish. 'interactive' and 'notebook' aren't exactly the same thing. I run interactively in line mode (in a terminal not in a notebook). You mention results, and yes, you can set the 'results' for tabular output to be either 'pandas' (the default), 'html' or 'text'. This controls how tabular results are returned to SASPy from SAS (table or datasets). The log output will be text, unless you use the colorLOG config setting in which case that becomes html. When using the submit*() methods, the results setting can affect the Listing (LST). If results is 'text' then the LST will be text. Again, I use this setting when running in a terminal session since HTML won't be rendered in line mode. Otherwise the LST will be returned as HTML (ODS output) from the submit methods.

Output from analytic methods often contains multiple outputs, some tabluar and others as plots/graphs/charts (these are ODS output which are HTML by default). These plots/graphs ... can't be returned as text, so those will always be the ODS output, which again, is HTML by default. But other results from them that aren't plots/graphs abide by the 'results' setting.

**Comment by srisa-work** [comment]
In practice my confusion is that I did not know there was a pandas output option.
How about the cshrc question?

Thanks

incidentally... would pandas also apply to line 64 of this? https://github.com/sassoftware/saspy/blob/main/saspy/sascfg.py

**Reply by tomweber-sas** [reply]
# 'output' = ['html5', 'html']

No, that's an option that I added for a user that was needing to use SAS 9.3 (which isn't supported) for a specific situation where they didn't have issue with the limitation of using html instead of html5. That has nothing to do with pandas.
